src/client/visualizations/bar-chart/improved-bar-chart/interactions/interaction-controller.tsx

```diff
@@ -0,0 +1,145 @@
+import { Essence } from "../../../../../common/models/essence/essence";
+import { FilterClause, FixedTimeFilterClause } from "../../../../../common/models/filter-clause/filter-clause";
+import { ConcreteSeries } from "../../../../../common/models/series/concrete-series";
+import { Binary, Unary } from "../../../../../common/utils/functional/functional";
+import { safeEquals } from "../../../../../common/utils/immutable-utils/immutable-utils";
+import { ScrollerPart } from "../../../../components/scroller/scroller";
+import { Highlight } from "../../../base-visualization/highlight";
+import { toPlywoodRange } from "../../../line-chart/interactions/highlight-clause";
+import { selectFirstSplitDatums } from "../../../line-chart/utils/dataset";
+import { BarChartLayout } from "../utils/layout";
+import { firstSplitRef } from "../utils/splits";
+import { DomainValue } from "../utils/x-domain";
+import { XScale } from "../utils/x-scale";
+import { createHighlight, createHover, equalInteractions, Hover, Interaction } from "./interaction";
+
+interface InteractionProps {
+  onClick?: (x: number, y: number, part: ScrollerPart) => void;
+  onMouseMove?: (x: number, y: number, part: ScrollerPart) => void;
+  onMouseLeave?: () => void;
+  onScroll?: (scrollTop: number, scrollLeft: number) => void;
+  interaction: Interaction | null;
+}
+
+interface InteractionControllerProps {
+  xScale: XScale;
+  essence: Essence;
+  dataset: Dataset;
+  layout: BarChartLayout;
+  children: Unary<InteractionProps, React.ReactNode>;
+  highlight?: Highlight;
+  saveHighlight: Binary<List<FilterClause>, string, void>;
+}
+
+interface InteractionControllerState {
+  hover?: Hover;
+  scrollLeft: number;
+  scrollTop: number;
+}
+
+export class InteractionController extends React.Component<InteractionControllerProps, InteractionControllerState> {
+
+  state: InteractionControllerState = { hover: null, scrollLeft: 0, scrollTop: 0 };
+
+  saveScroll = (scrollTop: number, scrollLeft: number) => {
+    this.setState({
+      hover: null,
+      scrollLeft,
+      scrollTop
+    });
+  };
+
+  saveHover = (x: number, y: number, part: ScrollerPart) => {
+    const { highlight } = this.props;
+    if (highlight) return;
+    const value = this.getValueFromEvent(x, part);
+    if (!TimeRange.isTimeRange(value)) return;
+    const series = this.getSeriesFromEvent(y, part);
+    const datum = this.findDatumByValue(value);
+    const hover = createHover(series.plywoodKey(), datum);
+    const { hover: oldHover } = this.state;
+    if (oldHover && equalInteractions(oldHover, hover)) return;
+    this.setState({ hover });
+  };
+
+  resetHover = () => {
+    const { hover } = this.state;
+    if (hover) {
+      this.setState({ hover: null });
+    }
+  };
+
+  handleClick = (x: number, y: number, part: ScrollerPart) => {
+    const value = this.getValueFromEvent(x, part);
+    if (!TimeRange.isTimeRange(value)) return;
```

Comment:
duplicated invariant check in saveHover and handleClick

Reply:
I don't follow. These are two different entry points. I don't see a way to abstract that further - I have method that calculates something and may fail and I need to check for failure.